### PR TITLE
Avoid deprecated findbugs annotations

### PR DESCRIPTION
**Pull request by olivergondza**
Move to `javax.annotations` where possible. Plus `SuppressWarnings` -> `SuppressFBWarnings`.

### CODE REVIEW

File: test/0/build.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (oleg-nenashev) [inline test/0/build.xml]: Uhm? Seems it has been occasionally committed

File: core/src/main/java/hudson/util/XStream2.java [hunk 2]
```diff
@@ -415,7 +415,7 @@ public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext co
 
         private PluginManager pm;
 
-        @SuppressWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE") // classOwnership checked for null so why does FB complain?
+        @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE") // classOwnership checked for null so why does FB complain?
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/util/XStream2.java]: Good practice is to add a clarification of suppression reasons. There is a field in the attribute for it

**Comment by KostyaSha** [comment]
:+1:

File: test/0/build.xml (hunk 1 above)
Comment (olivergondza) [inline test/0/build.xml]: Indeed.

File: core/src/main/java/hudson/util/XStream2.java (hunk 2 above)
Comment (olivergondza) [inline core/src/main/java/hudson/util/XStream2.java]: I agree, though it should have been done when the annotations was introduced initially. I do not think the inline comment is a good `justification` btw.

**Comment by oleg-nenashev** [comment]
:+1:

**Comment by jtnord** [comment]
:-1: we should not be using JSR305 `javax.annotations`  - see findbugs/issues/88 for the reasons why.

@stephenc is updating his clean room port of findbugs (sans deprecation flags). so we have an implementation we can use that has a nice licence, and is working on a PR to get rid of all the illegal annotation use in the project. 

At some point I will also send a PR to findbugs to get rid of their self induced mess.

However changing `@SuppressWarnings` to `@SuppressFBWarnings` gets my vote.

**Comment by KostyaSha** [comment]
@jtnord what issues do you have with license?

**Comment by oleg-nenashev** [comment]
@jtnord 
javax.annotations are everywhere in Jenkins core and plugins. I agree with your point, but I don't feel it's a blocker for this PR. 

If we want to change it, a major bulk rework is required.

**Comment by jtnord** [comment]
@KostyaSha its to do with the oracle java licence agreement and using the `java`/`javax` namespace to include things that are not official (JSR-305 is not official - it never produced any offical API - what exists is a point in time unofficial thing that was in someones head) - so this hits the docker images more than the wars - but still.  And we should not be distributing the annotations - however accidents happen, and as we have a perfectly good alternative (apart from the deprecations...)

@oleg-nenashev I believe @stephenc has this rework on his radar for a rainy day - so yes this is a large amount of work.  My point is lets not make this worse in the interim - softly softly catchy monkey!

**Comment by stephenc** [comment]
IMHO the correct thing is not to introduce more `javax.annotation` usage. When I get the time - hopefully before 2.0 - I will be creating a PR to make the change of **every one** of the `javax.annotation` annotations with the corresponding non-`javax.annotation` annotation.

The PR will be more annoying if I have classes that have both annotations included as there will be the FQCN on the annotations to resolve the conflict. Here is my simple rule to make the PR easier:

* If the class already has an import for the specific `javax.annotation` annotation, leave that import alone and use it.
* If the class already has an import for the specific `edu.umd.cs.findbugs` annotation, leave that import alone and use it.

So for example if a class has the following imports:

```
import javax.annotation.CheckForNull;
import edu.umd.cs.findbugs.Nullable;
```

Then all the `@CheckForNull` annotations in that class should be consolidated to `javax.annotation.CheckForNull` and all the `@Nullable` annotations in that class should be consolidated to `edu.umd.cs.findbugs.Nullable` i.e. replace and `@edu.umd.cs.findbugs.CheckForNull` annotations with `@CheckForNull` and replace any `@javax.annotation.Nullable` annotations with `@Nullable`.

If adding an annotation that has not been used, follow majority rule based on the existing imports. In the event of a tie, `edu.umd.cs.findbugs` wins.

**Comment by olivergondza** [comment]
I do not feel strongly w.r.t. the license thing. There are so many projects using it and we used to prefer javax over findbugs in Jenkins for as long as I remember.

However, this PR gets rid of all findbugs annotations leaving only `SuppressFBWarnings` (that does not have an alternative in `javax.annotation`). So if we decide to switch to some other annotations set in the future, it should be easier to bulk update the sources with this change in.

**Comment by stephenc** [comment]
@olivergondza so if we (i.e. the Jenkins project) were only shipping a WAR file, I would mostly agree with your "don't care" position.

The problem became critical as soon as the Jenkins project started shipping anything that bundles a JVM... https://jenkins-ci.org/blog/2014/08/12/official-jenkins-lts-docker-image/ is one example of such a thing.

In order to distribute a JVM you must comply with the Oracle redistribution terms. Specifically:

> F. JAVA TECHNOLOGY RESTRICTIONS. You may not create, modify, or change the behavior of, or authorize your licensees to create, modify, or change the behavior of, classes, interfaces, or subpackages that are in any way identified as "java", "javax", "sun", “oracle” or similar convention as specified by Oracle in any naming convention designation.

Now if the [JSR 305](https://jcp.org/en/jsr/detail?id=305) page publishes javadoc for these classes, or publishes a jar file with these classes, by virtue of that being from the official JSR process, we would be OK... but because there is no JSR page anywhere with these annotations, we are therefore breaking the redistribution terms for a JRE (we'd be ok if we were redistributing Apache Harmony as that did not claim to be a "Java"... similarly if redistributing a "Davilk") but because we are redistributing something that claims to be a compliant Java™ JVM in the docker image... we (The Jenkins Project) are technically redistributing that without a license to do so.

Hence, we need to stop fairly soonish

**Comment by stephenc** [comment]
OT, but one of the reasons why our installers do not vendorize a JVM (which is the Oracle recommended way to ship larger JVM apps) is because we would then have to worry about things like the [Oracle Java binary licence](http://www.oracle.com/technetwork/java/javase/terms/license/index.html)

Of course if we vendorized the JVM into our installers, we'd also have to worry about JVM vulnerabilities, etc... and that is a bigger worry than the redistribution terms

**Comment by jtnord** [comment]
> OT, but one of the reasons why our installers do not vendorize a JVM

I had to rain on your parade but we do...  Its the windows installer to "make things simple"<sup>:tm:</sup>

Don't shoot the messenger... [JENKINS-31659](https://issues.jenkins-ci.org/browse/JENKINS-31659)

**Comment by stephenc** [comment]
@jtnord well then there's another binary besides docker... at least until JENKINS-31659 from my reading of it

**Comment by oleg-nenashev** [comment]
Anyway, it didn't get into before Jenkins 2

**Comment by oleg-nenashev** [comment]
So we screwed up the useful change in order to implement something bigger :(
@olivergondza What would be the next steps? Should it be just closed?

**Comment by olivergondza** [comment]
I still think this is worth doing. 

> However, this PR gets rid of all findbugs annotations leaving only SuppressFBWarnings (that does not have an alternative in javax.annotation). So if we decide to switch to some other annotations set in the future, it should be easier to bulk update the sources with this change in.

**Comment by oleg-nenashev** [comment]
re-:+1:

**Comment by stephenc** [comment]
re :-1:

**Comment by olivergondza** [comment]
@stephenc, @jtnord, do you care to explain? If we are in fact affected by the licensing problem, adding 20 more problematic annotations makes no difference if we use hundreds on them already. And as I said, unifying it now make it easier to switch to fixed version in the future. Plus, I do not see anybody working on removing/replacing existing usage of `javax.annotation` to *resolve* the problem you have coined.

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers To there is 3 vs 2. Some additional votes would be useful

**Comment by daniel-beck** [comment]
@stephenc You're effectively blocking this PR saying that it would make your cleanup more difficult, but haven't followed up in half a year (when originally it was supposed to take a month at most, tied to 2.0 release). It's really time you finally implemented this, or stop discouraging this change.

**Comment by stephenc** [comment]
Look if you want to switch to the `javax` packages then you need to:

1. Stop shipping a JRE in the windows installer
2. Stop shipping a JRE in the docker image

Do both of those things and I will reverse my vote... or stop using `javax` packages which have not been released as part of an official JSR release process. (The terms of redistribution of a JRE explicitly forbid redistribution with any code that is in either `java` or `javax` that has not been released by a JSR... we can be forgiven for leaving the status quo... but moving in the wrong direction is willful disregard)

Or you can ignore my vote. @jtnord and I are not seeing much progress from the findbugs people...

**Comment by KostyaSha** [comment]
Using or shipping? FB is only compile time dependency and annotation existence is optional on binary level AFAIR.

**Comment by stephenc** [comment]
Well last I checked we bundle the annotations... would need legal advice to differentiate shipping vs using annotations within the JRE redistribution T&C

**Comment by oleg-nenashev** [comment]
Merging since it does not make the situation worse && @stephenc accepts merge && @jtnord has not responded. I definitely would like to see a solution with less legal concerns, but it's what we have now.

**Pull request merged**